go/rpcclientlib/viewing_key_client.go

```diff
@@ -216,14 +221,40 @@ func (c *ViewingKeyClient) addFromAddressToCallParamsIfMissing(method string, ar
 		}
 	}
 
+	// We only modify `eth_call` requests where the `from` field is not set.
 	if callParams[reqJSONKeyFrom] != nil {
-		// We only modify `eth_call` requests where the `from` field is not set.
 		return args, nil
 	}
-	callParams[reqJSONKeyFrom] = c.viewingKeyAddr
 
-	args[0] = callParams
+	// We ensure that the `data` field is present.
+	if callParams[reqJSONKeyData] == nil {
+		return nil, fmt.Errorf("eth_call request did not have its `from` or its `data` field set. Aborting " +
+			"request as it will not be possible to encrypt the response")
+	}
+	data, ok := callParams[reqJSONKeyData].(string)
+	if !ok {
+		return nil, fmt.Errorf("eth_call request's `data` field was not of the expected type `string`")
+	}
 
+	// We iterate over the `data` field, looking for an argument that matches the viewing key address. If we find one,
+	// we set the `from` field to that address.
+	data = data[10:] // We remove the leading "0x" (1 bytes/2 chars) and the method ID (4 bytes/8 chars).
+	for i := 0; i < len(data); i += ethCallPaddedArgLen {
```

Comment:
Maybe it would be easier to follow it if was formatted to loop over args like a `while hasNext`, e.g.

```
for arg := getNextArg(data); nextArg != nil; {
   // check arg for address
   ...
    arg = getNextArg(data)
}
```

Reply:
I'm splitting it into steps - one to extract all the args, and one to process each arg in turn.